c/meterpreter/source/extensions/stdapi/server/sys/config/config.c

```diff
@@ -590,14 +590,96 @@ DWORD add_windows_os_version(Packet** packet)
 		{
 			if (v.dwMinorVersion == 0)
 			{
-				if (v.dwBuildNumber < 17763) {
-					osName = v.wProductType == VER_NT_WORKSTATION ? "Windows 10" : "Windows Server 2016";
-				} else if (v.dwBuildNumber < 20348) {
-					osName = v.wProductType == VER_NT_WORKSTATION ? "Windows 10" : "Windows Server 2019";
-				} else if (v.dwBuildNumber < 22000) {
-					osName = v.wProductType == VER_NT_WORKSTATION ? "Windows 10" : "Windows Server 2022";
-				} else {
-					osName = v.wProductType == VER_NT_WORKSTATION ? "Windows 11" : "Windows Server 2022";
+				if (v.wProductType == VER_NT_WORKSTATION) {
+					if (v.dwBuildNumber < 10586) {
+						osName = "Windows 10";
+					}
+					else if (v.dwBuildNumber < 14393) {
+						osName = "Windows 10 1511";
+					}
+					else if (v.dwBuildNumber < 15063) {
+						osName = "Windows 10 1607";
+					}
+					else if (v.dwBuildNumber < 16299) {
+						osName = "Windows 10 1703";
+					}
+					else if (v.dwBuildNumber < 17134) {
+						osName = "Windows 10 1709";
+					}
+					else if (v.dwBuildNumber < 17763) {
+						osName = "Windows 10 1803";
+					}
+					else if (v.dwBuildNumber < 18362) {
+						osName = "Windows 10 1809";
+					}
+					else if (v.dwBuildNumber < 18363) {
+						osName = "Windows 10 1903";
+					}
+					else if (v.dwBuildNumber < 19041) {
+						osName = "Windows 10 1909";
+					}
+					else if (v.dwBuildNumber < 19042) {
+						osName = "Windows 10 2004";
+					}
+					else if (v.dwBuildNumber < 19043) {
+						osName = "Windows 10 20H2";
+					}
+					else if (v.dwBuildNumber < 19044) {
+						osName = "Windows 10 21H1";
+					}
+					else if (v.dwBuildNumber < 19045) {
+						osName = "Windows 10 21H2";
+					}
+					else if (v.dwBuildNumber < 22000) {
+						osName = "Windows 10 22H2+";
+					}
+					else if (v.dwBuildNumber < 22621) {
+						osName = "Windows 11 21H2";
+					}
+					else if (v.dwBuildNumber < 22631) {
+						osName = "Windows 11 22H2";
+					}
+					else if (v.dwBuildNumber < 26100) {
+						osName = "Windows 11 23H2";
+					}
+					else {
+						osName = "Windows 11 24H2+";
+					}
+				}
+				else {
+					if (v.dwBuildNumber < 16299) {
+						osName = "Windows Server 2016";
+					}
+					else if (v.dwBuildNumber < 17134) {
+						osName = "Windows Server 1709";
+					}
+					else if (v.dwBuildNumber < 17763) {
+						osName = "Windows Server 1803";
+					}
+					else if (v.dwBuildNumber < 18362) {
+						osName = "Windows Server 2019/1809";
+					}

```

Comment:
Sorry, what I meant in my recommendation was that this should be:
```suggestion
					else if (v.dwBuildNumber < 18362) {
						osName = "Windows Server 2019 1809";
					}

```

And that the pattern should be applied to each, so `Windows Server 2004` isn't confused with a Windows server version released in the year 2004 (because there wasn't any). This would follow the pattern of Windows 10/11 above where the product name is proceeded by the version.